channels/pkg/channels/apply.go

```diff
@@ -17,53 +17,114 @@ limitations under the License.
+		return fmt.Errorf("failed to parse objects: %w", err)
 	}
 
-	defer func() {
-		if err := os.RemoveAll(tmpDir); err != nil {
-			klog.Warningf("error deleting temp dir %q: %v", tmpDir, err)
+	objectsByKind := make(map[schema.GroupKind][]*kubemanifest.Object)
+	for _, object := range objects {
+		gv, err := schema.ParseGroupVersion(object.APIVersion())
+		if err != nil || gv.Version == "" {
+			return fmt.Errorf("failed to parse apiVersion %q", object.APIVersion())
+		}
+		kind := object.Kind()
+		if kind == "" {
+			return fmt.Errorf("failed to find kind in object")
 		}
-	}()
 
-	localManifestFile := path.Join(tmpDir, "manifest.yaml")
-	if err := os.WriteFile(localManifestFile, data, 0o600); err != nil {
-		return fmt.Errorf("error writing temp file: %v", err)
+		gvk := gv.WithKind(kind)
+		gk := gvk.GroupKind()
+		objectsByKind[gk] = append(objectsByKind[gk], object)
 	}
 
-	_, err = execKubectl("replace", "-f", localManifestFile, "--force", "--field-manager=kops")
-	return err
+	for gk := range objectsByKind {
+		if err := p.applyObjectsOfKind(ctx, gk, objectsByKind[gk]); err != nil {
+			return fmt.Errorf("failed to apply objects of kind %s: %w", gk, err)
+		}
+	}
+	return nil
 }
 
-func execKubectl(args ...string) (string, error) {
-	kubectlPath := "kubectl" // Assume in PATH
-	cmd := exec.Command(kubectlPath, args...)
-	env := os.Environ()
-	cmd.Env = env
+func (p *Applier) applyObjectsOfKind(ctx context.Context, gk schema.GroupKind, expectedObjects []*kubemanifest.Object) error {
+	klog.V(2).Infof("applying objects of kind: %v", gk)
+
+	restMapping, err := p.RESTMapper.RESTMapping(gk)
+	if err != nil {
+		return fmt.Errorf("unable to find resource for %s: %w", gk, err)
+	}
+
+	gvr := restMapping.Resource
 
-	human := strings.Join(cmd.Args, " ")
-	klog.V(2).Infof("Running command: %s", human)
-	output, err := cmd.CombinedOutput()
+	baseResource := p.Client.Resource(gvr)
+
+	actualObjects, err := baseResource.List(ctx, v1.ListOptions{})
 	if err != nil {
-		klog.Infof("error running %s", human)
-		klog.Info(string(output))
-		return string(output), fmt.Errorf("error running kubectl: %v", err)
+		return fmt.Errorf("error listing objects: %w", err)
+	}
+	if err := p.applyObjects(ctx, gvr, actualObjects, expectedObjects); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Applier) applyObjects(ctx context.Context, gvr schema.GroupVersionResource, actualObjects *unstructured.UnstructuredList, expectedObjects []*kubemanifest.Object) error {
+	actualMap := make(map[string]unstructured.Unstructured)
+	for _, actualObject := range actualObjects.Items {
+		key := actualObject.GetNamespace() + "/" + actualObject.GetName()
+		actualMap[key] = actualObject
+	}
+
+	for _, expectedObjects := range expectedObjects {
+		name := expectedObjects.GetName()
+		namespace := expectedObjects.GetNamespace()
+		key := namespace + "/" + name
+
+		var resource dynamic.ResourceInterface
+		if namespace != "" {
```

Comment:
Nit: If we pass in the restMapping, we can know if this object is namespaced or not (and raise an appropriate error/warning or default the namespace, whatever we want to do)